File: framework/src/main/java/com/thornbirds/framework/adapter/SingleChooser.java
package com.thornbirds.framework.adapter;

import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewGroup;

import static android.view.View.NO_ID;

/**
 * This class defines a helper for choosing one item from several view.
 *
 * @author YangLi yanglijd@gmail.com
 */
public class SingleChooser implements View.OnClickListener {
    private static final String TAG = "SingleChooser";

    private IChooserListener mListener;

    protected ViewGroup mViewGroup;
    protected View mSelectedView;

    private <T extends View> T $(@IdRes int resId) {
        return mViewGroup != null ? (T) mViewGroup.findViewById(resId) : null;
    }

    @Override
    public void onClick(View view) {
        boolean ret = setSelection(view);
        if (ret && mListener != null) {
            mListener.onItemSelected(mSelectedView);
        }
    }

    public SingleChooser(IChooserListener listener) {
        mListener = listener;
    }

    public boolean setSelection(View view) {
        if (mSelectedView != null && mSelectedView == view) {
            return false;
        }
        if (mSelectedView != null) {
            mSelectedView.setSelected(false);
        }
        mSelectedView = view;
        mSelectedView.setSelected(true);
        return true;
    }

    public void setSelection(@IdRes int selectedId) {
        View view = $(selectedId);
        if (view != null) {
            setSelection(view);
        }
    }

    public void setup(@NonNull ViewGroup viewGroup, @IdRes int selectedId) {
        if (mViewGroup != null && mViewGroup != viewGroup) {
            reset();
        }
        mViewGroup = viewGroup;

        int childCount = mViewGroup.getChildCount();
        for (int i = 0; i < childCount; ++i) {
            View view = mViewGroup.getChildAt(i);
            int id = view.getId();
            if (id == NO_ID) {
                continue;
            }
            if (id == selectedId) {
                mSelectedView = view;
                mSelectedView.setSelected(true);
            }
            view.setOnClickListener(this);
        }
    }

    public void reset() {
        if (mViewGroup == null) {
            return;
        }
        int childCount = mViewGroup.getChildCount();
        for (int i = 0; i < childCount; ++i) {
            View view = mViewGroup.getChildAt(i);
            view.setOnClickListener(null);
        }
        mViewGroup = null;
        mSelectedView = null;
    }

    public interface IChooserListener {

        void onItemSelected(View view);

    }
}

File: framework/src/main/java/com/thornbirds/framework/adapter/SingleChooser.kt
package com.thornbirds.framework.adapter

import android.view.View
import android.view.ViewGroup
import androidx.annotation.IdRes

/**
 * This class defines a helper for choosing one item from several view.
 *
 * @author YangLi yanglijd@gmail.com
 */
class SingleChooser(
    private val mListener: IChooserListener?
) : View.OnClickListener {

    private var mViewGroup: ViewGroup? = null
    private var mSelectedView: View? = null

    private fun <T : View?> findViewById(@IdRes resId: Int): T? {
        return mViewGroup?.findViewById(resId)
    }

    override fun onClick(view: View) {
        val ret = setSelection(view)
        if (ret && mListener != null) {
            mListener.onItemSelected(view)
        }
    }

    fun setSelection(view: View): Boolean {
        val selectedView = mSelectedView
        if (selectedView != null && selectedView === view) {
            return false
        }
        if (selectedView != null) {
            selectedView.isSelected = false
        }
        mSelectedView = view.also {
            it.isSelected = true
        }
        return true
    }

    fun setSelection(@IdRes selectedId: Int) {
        findViewById<View>(selectedId)?.let {
            setSelection(it)
        }
    }

    fun setup(viewGroup: ViewGroup, @IdRes selectedId: Int) {
        if (mViewGroup != null && mViewGroup != viewGroup) {
            reset()
        }
        mViewGroup = viewGroup
        val childCount = viewGroup.childCount
        for (i in 0 until childCount) {
            val view = viewGroup.getChildAt(i)
            val id = view.id
            if (id == View.NO_ID) {
                continue
            }
            if (id == selectedId) {
                setSelection(view)
            }
            view.setOnClickListener(this)
        }
    }

    fun reset() {
        mViewGroup?.let { viewGroup ->
            val childCount = viewGroup.childCount
            for (i in 0 until childCount) {
                val view = viewGroup.getChildAt(i)
                view.setOnClickListener(null)
            }
            mViewGroup = null
            mSelectedView = null
        }
    }

    interface IChooserListener {
        fun onItemSelected(view: View)
    }
}